drivers/disk/flashdisk.c

```diff
@@ -52,38 +49,115 @@ static int disk_flash_access_status(struct disk_info *disk)
 
 static int disk_flash_access_init(struct disk_info *disk)
 {
-	if (flash_dev) {
-		return 0;
+	struct flashdisk_data *ctx;
+	const struct flash_area *fap;
+	int rc;
+	struct flash_pages_info page;
+	off_t offset;
+
+	ctx = CONTAINER_OF(disk, struct flashdisk_data, info);
+
+	rc = flash_area_open(ctx->area_id, &fap);
+	if (rc < 0) {
+		LOG_ERR("Flash area %u open error %d", ctx->area_id, rc);
+		return rc;
+	}
+
+	disk->dev = flash_area_get_device(fap);
+
+	rc = flash_get_page_info_by_offs(disk->dev, ctx->offset, &page);
+	if (rc < 0) {
+		LOG_ERR("Error %d while getting page info", rc);
+		flash_area_close(fap);
+		return rc;
+	}
+
+	ctx->page_size = page.size;
+	LOG_INF("Initialize device %s", disk->name);
+	LOG_INF("offset %lx, sector size %zu, page size %zu, volume size %zu",
+		(long)ctx->offset, ctx->sector_size, ctx->page_size, ctx->size);
+
+	if (ctx->offset != page.start_offset) {
+		LOG_ERR("Partition does not start at page boundary");
+		flash_area_close(fap);
+		return -EINVAL;
+	}
+
+	offset = ctx->offset + page.size;
+	while (offset < ctx->offset + ctx->size) {
+		rc = flash_get_page_info_by_offs(disk->dev, offset, &page);
+		if (rc < 0) {
+			LOG_ERR("Error %d while getting page info", rc);
+			flash_area_close(fap);
+			return rc;
+		}
+		if (page.size != ctx->page_size) {
+			LOG_ERR("Non-uniform page size is not supported");
+			flash_area_close(fap);
+			return rc;
+		}
+		offset += page.size;
+	}
```

Comment:
This breaks ability to have RO FAT FS on non page uniform devices.

Reply:
I'll just remove the constraints check commit then. The write on non-uniform partitions never worked before and this PR won't change it, but removing the constraint check will make the RO continue to work.